File: lintcode/golang/1378_minimum_string_array_coverage.go
/**
 * @param tagList: The tag list.
 * @param allTags: All the tags.
 * @return: Return the answer
 */

import "fmt"

func getMinimumStringArray (tagList []string, allTags []string) int {
    // Write your code here
    tagListDic := make(map[string]bool)
    foundTagsDic := make(map[string]int)

    n := len(tagList)
    m := len(allTags)

    for i := 0; i < n; i++ {
        tagListDic[tagList[i]] = true
        foundTagsDic[tagList[i]] = 0
    }
    fmt.Println("tagdic: ", tagListDic)
    fmt.Println("china is in: ", tagListDic["china"])

    ret := m+1
    cnt := 0
    left, right := 0, -1

    for right < m {
        fmt.Println("left: ", left, "right: ", right)
        fmt.Println("left word: ", allTags[left])
        if cnt < n {
            right++
            fmt.Println("right word: ", allTags[right], "tagdic: ", tagListDic[allTags[right]])
            if right == m {
                break
            }
            if tagListDic[allTags[right]] == true {
                foundTagsDic[allTags[right]]++
                if foundTagsDic[allTags[right]] == 1 {
                    cnt++
                }
            }
            fmt.Println("foundTagDic: ", foundTagsDic, "cnt: ", cnt)
        } else {
            if tagListDic[allTags[left]] == true {
                foundTagsDic[allTags[left]]--
                if foundTagsDic[allTags[left]] == 0 {
                    cnt--
                }
            }
            left++
        }
        if cnt == n {
            ret = min(ret, right-left+1)
        }
    }

    if ret == m+1 {
        return -1
    }

    return ret
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: lintcode/python/1378_minimum_string_array_coverage.py
class Solution:
    """
    @param tagList: The tag list.
    @param allTags: All the tags.
    @return: Return the answer
    """
    def getMinimumStringArray(self, tagList, allTags):
        # Write your code here
        mp1, mp2 = {}, {}
        for i in tagList:
            mp1[i] = 1
            mp2[i] = 0
        l , r = 0, -1
        n = len(allTags)
        m = len(tagList)
        cnt = 0
        ans = n + 1
        while(r < n):
            if(cnt < m):
                r += 1
                if(r == n):
                    break
                if(allTags[r]  in mp1):
                    mp2[allTags[r]] += 1
                    if(mp2[allTags[r]] == 1):
                        cnt += 1
            else:
                if(allTags[l] in mp1):
                    mp2[allTags[l]] -= 1
                    if(mp2[allTags[l]] == 0):
                        cnt -= 1
                l += 1
            if(cnt == m):
                ans = min(ans, r - l + 1)
        if(ans == n + 1):
            ans = -1
        return ans
